In [ ]:
# Load the dataset
import pandas as pd

df = pd.read_csv("nintendo_games.csv")
df.head()

,gender,rank type,id,name,rankScore,sinopsis,tags,rank,price,Category,Players,Published by,Europe release,US release,Japan release,Developed by,Series,Collection
0,Adventure,ranked,4423,A Short Hike,9.34,A Short Hike is an adventure game for the Nint...,"Tags - Adventure, Exploration, Open world, Col...",16 / 4512,6.29,Adventure,1,Whippoorwill,18th Aug 2020,18th Aug 2020,24th Sep 2020,NaN,NaN,NaN
1,Adventure,ranked,3449,Kentucky Route Zero: TV Edition,9.10,Kentucky Route Zero: TV Edition is an adventur...,"Tags - Adventure, Story choices, Emotional, Me...",49 / 4512,20.99,Adventure,1,Annapurna Interactive,28th Jan 2020,28th Jan 2020,28th Jan 2020,Cardboard Computer,NaN,NaN
2,Adventure,ranked,8981,The Oregon Trail,9.07,The Oregon Trail is an adventure game for the ...,"Tags - Adventure, Survival, Story choices, Wes...",54 / 4512,26.99,Adventure,NaN,Gameloft,14th Nov 2022,NaN,NaN,NaN,NaN,NaN
3,Adventure,ranked,7399,13 Sentinels: Aegis Rim,8.92,13 Sentinels: Aegis Rim is an adventure game f...,"Tags - Adventure, RPG, Strategy, Visual novel,...",97 / 4512,49.99,Adventure,1,Atlus Sega,12th Apr 2022,NaN,NaN,Vanillaware,NaN,NaN
4,Adventure,ranked,2829,AI: The Somnium Files,8.86,AI: The Somnium Files is an adventure game for...,"Tags - Adventure, Point-and-click, Visual nove...",116 / 4512,35.99,Adventure,1,Spike Chunsoft,20th Sep 2019,17th Sep 2019,19th Sep 2019,Spike Chunsoft,NaN,NaN


Load the dataset of Nintendo Switch games from: https://www.kaggle.com/datasets/pedroaltobelli/nintendo-switch-games.

In [ ]:
# Show dataset info and shape
display(df.info())
display(df.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13152 entries, 0 to 13151
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   gender          13152 non-null  object 
 1   rank type       13152 non-null  object 
 2   id              13152 non-null  int64  
 3   name            13152 non-null  object 
 4   rankScore       13152 non-null  object 
 5   sinopsis        13152 non-null  object 
 6   tags            13152 non-null  object 
 7   rank            13152 non-null  object 
 8   price           13145 non-null  float64
 9   Category        13152 non-null  object 
 10  Players         8445 non-null   object 
 11  Published by    13152 non-null  object 
 12  Europe release  13152 non-null  object 
 13  US release      3734 non-null   object 
 14  Japan release   2183 non-null   object 
 15  Developed by    2629 non-null   object 
 16  Series          843 non-null    object 
 17  Collection      635 non-null   

None

(13152, 18)

This step displays metadata of the dataset.

In [ ]:
# Delete unused column
to_drop = ['id', 'gender', 'Europe release', 'Japan release', 'US release']
games_df = df.drop(columns=to_drop)
games_df.head()

,rank type,name,rankScore,sinopsis,tags,rank,price,Category,Players,Published by,Developed by,Series,Collection
0,ranked,A Short Hike,9.34,A Short Hike is an adventure game for the Nint...,"Tags - Adventure, Exploration, Open world, Col...",16 / 4512,6.29,Adventure,1,Whippoorwill,NaN,NaN,NaN
1,ranked,Kentucky Route Zero: TV Edition,9.10,Kentucky Route Zero: TV Edition is an adventur...,"Tags - Adventure, Story choices, Emotional, Me...",49 / 4512,20.99,Adventure,1,Annapurna Interactive,Cardboard Computer,NaN,NaN
2,ranked,The Oregon Trail,9.07,The Oregon Trail is an adventure game for the ...,"Tags - Adventure, Survival, Story choices, Wes...",54 / 4512,26.99,Adventure,NaN,Gameloft,NaN,NaN,NaN
3,ranked,13 Sentinels: Aegis Rim,8.92,13 Sentinels: Aegis Rim is an adventure game f...,"Tags - Adventure, RPG, Strategy, Visual novel,...",97 / 4512,49.99,Adventure,1,Atlus Sega,Vanillaware,NaN,NaN
4,ranked,AI: The Somnium Files,8.86,AI: The Somnium Files is an adventure game for...,"Tags - Adventure, Point-and-click, Visual nove...",116 / 4512,35.99,Adventure,1,Spike Chunsoft,Spike Chunsoft,NaN,NaN


Remove unused columns. Unused columns does not contains important data for this project.

In [ ]:
# Check for duplicated data
display(games_df.duplicated().sum())

np.int64(0)

Check for any duplicated data inside games_df. It found 0 data are duplicated within.

In [ ]:
# Check for empty / Nan data
display(games_df.isna().sum())

,0
rank type,0
name,0
rankScore,0
sinopsis,0
tags,0
rank,0
price,7
Category,0
Players,4707
Published by,0


Check for any empty/NaN data within. It shows there are still any data that may have empty data on columns "price", "Players", "Developed by", "Series", and "Collection".

In [ ]:
# Remove any empty/NaN and unreliable data
to_drop = ['Players', 'Developed by', 'Series', 'Collection']
games_df_1 = games_df.drop(columns=to_drop)
games_df_2 = games_df_1.dropna(subset=['price'])
display(games_df_2.head())
display(games_df_2.isna().sum())
display(games_df_2.info())

,rank type,name,rankScore,sinopsis,tags,rank,price,Category,Published by
0,ranked,A Short Hike,9.34,A Short Hike is an adventure game for the Nint...,"Tags - Adventure, Exploration, Open world, Col...",16 / 4512,6.29,Adventure,Whippoorwill
1,ranked,Kentucky Route Zero: TV Edition,9.10,Kentucky Route Zero: TV Edition is an adventur...,"Tags - Adventure, Story choices, Emotional, Me...",49 / 4512,20.99,Adventure,Annapurna Interactive
2,ranked,The Oregon Trail,9.07,The Oregon Trail is an adventure game for the ...,"Tags - Adventure, Survival, Story choices, Wes...",54 / 4512,26.99,Adventure,Gameloft
3,ranked,13 Sentinels: Aegis Rim,8.92,13 Sentinels: Aegis Rim is an adventure game f...,"Tags - Adventure, RPG, Strategy, Visual novel,...",97 / 4512,49.99,Adventure,Atlus Sega
4,ranked,AI: The Somnium Files,8.86,AI: The Somnium Files is an adventure game for...,"Tags - Adventure, Point-and-click, Visual nove...",116 / 4512,35.99,Adventure,Spike Chunsoft


,0
rank type,0
name,0
rankScore,0
sinopsis,0
tags,0
rank,0
price,0
Category,0
Published by,0


<class 'pandas.core.frame.DataFrame'>
Index: 13145 entries, 0 to 13151
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rank type     13145 non-null  object 
 1   name          13145 non-null  object 
 2   rankScore     13145 non-null  object 
 3   sinopsis      13145 non-null  object 
 4   tags          13145 non-null  object 
 5   rank          13145 non-null  object 
 6   price         13145 non-null  float64
 7   Category      13145 non-null  object 
 8   Published by  13145 non-null  object 
dtypes: float64(1), object(8)
memory usage: 1.0+ MB


None

Columns "Players", "Developed by", "Series", and "Collection" are need to be deleted because it may unreliable because of the empty data are huge on them.

On column 'price', we just need to delete the data which has empty/NaN price only.

Now, the dataset doesnt contain any empty/NaN and unreliable data.

In [ ]:
# Check for "no tags" data
sum(games_df_2["tags"].apply(lambda x: isinstance(x, str) and "no tags" in x.lower()))

6325

This step checks for any game that has no tags on its tags column. It shows there are 6325 games that has no tags on its tags column.

In [ ]:
# Drop the data which have "no tags"
games_df_2 = games_df_2[
    ~games_df_2["tags"].str.lower().str.contains("no tags", na=False)
].reset_index(drop=True)
sum(games_df_2["tags"].apply(lambda x: isinstance(x, str) and "no tags" in x.lower()))

0

The data which have "no tags" on its tags column are being deleted because the column "tags" is importan feature.

In [ ]:
# Define column names
COL_APP   = "name"
COL_GENRE = "tags"
COL_DESC  = "sinopsis"
COL_CAT   = "Category"
COL_RATING = "rankScore"
COL_ISRANKED = "rank type"
COL_RANK = "rank"
COL_PRICE = "price"
COL_PUBLISHER = "Published by"


# Check existing columns
games_df_2.columns

Index(['rank type', 'name', 'rankScore', 'sinopsis', 'tags', 'rank', 'price',
       'Category', 'Published by'],
      dtype='object')

Standardized column mappings to ensure consistent access to key features in the dataset, such as app name, genre, description, category, rating, etc. Since different datasets may use different column names, these variables allow the code to remain flexible and reusable

In [ ]:
# Prepare the dataset
import re
import numpy as np
import pandas as pd

# -------------------------
# Helpers
# -------------------------
def clean_text(s):
    """Lowercase + normalize whitespace. NaN -> ''."""
    if pd.isna(s):
        return ""
    s = str(s).lower()
    s = re.sub(r"\s+", " ", s)
    return s.strip()

def parse_rank(s):
    """
    Parse '16 / 4512' -> (16, 4512).
    Returns (np.nan, np.nan) if can't parse.
    """
    if pd.isna(s):
        return (np.nan, np.nan)
    m = re.match(r"^\s*(\d+)\s*/\s*(\d+)\s*$", str(s))
    if not m:
        return (np.nan, np.nan)
    return (int(m.group(1)), int(m.group(2)))

def price_bucket(p):
    """Bucket price for content-based features."""
    if pd.isna(p):
        return "unknown_price"
    try:
        p = float(p)
    except Exception:
        return "unknown_price"
    if p == 0:
        return "free"
    elif p < 10:
        return "budget"
    elif p < 30:
        return "mid"
    else:
        return "premium"

games_df_3 = games_df_2.copy()

# -------------------------
# Validate columns
# -------------------------
required_cols = [COL_APP, COL_GENRE]
missing = [c for c in required_cols if c not in games_df_3.columns]
if missing:
    raise ValueError(f"Missing required columns: {missing}")

# Ensure optional columns exist
optional_cols = [COL_DESC, COL_CAT, COL_PUBLISHER, COL_ISRANKED, COL_RANK, COL_PRICE, COL_RATING]
for c in optional_cols:
    if c not in games_df_3.columns:
        games_df_3[c] = np.nan

# -------------------------
# Basic cleanup / typing
# -------------------------
games_df_3[COL_APP] = games_df_3[COL_APP].astype(str).str.strip()
games_df_3[COL_GENRE] = games_df_3[COL_GENRE].fillna("").astype(str)
games_df_3[COL_DESC] = games_df_3[COL_DESC].fillna("").astype(str)
games_df_3[COL_CAT] = games_df_3[COL_CAT].fillna("").astype(str)
games_df_3[COL_PUBLISHER] = games_df_3[COL_PUBLISHER].fillna("").astype(str)
games_df_3[COL_ISRANKED] = games_df_3[COL_ISRANKED].fillna("").astype(str)
games_df_3[COL_RANK] = games_df_3[COL_RANK].fillna("").astype(str)

# Make numeric where appropriate
games_df_3[COL_PRICE] = pd.to_numeric(games_df_3[COL_PRICE], errors="coerce")
games_df_3[COL_RATING] = pd.to_numeric(games_df_3[COL_RATING], errors="coerce")
games_df_3[COL_RATING] = games_df_3[COL_RATING].fillna(0)

# Remove duplicates by game name
games_df_3 = games_df_3.drop_duplicates(subset=[COL_APP]).reset_index(drop=True)

# -------------------------
# Parse rank column into numeric features
# -------------------------
rank_parsed = games_df_3[COL_RANK].apply(parse_rank)
games_df_3["rank_pos"] = rank_parsed.apply(lambda x: x[0])
games_df_3["rank_total"] = rank_parsed.apply(lambda x: x[1])

# Optional: normalize rank position into a score (higher is better)
# If rank_pos=1 and rank_total=4512 -> score close to 1.0
games_df_3["rank_norm"] = 1 - (games_df_3["rank_pos"] - 1) / (games_df_3["rank_total"] - 1)
games_df_3.loc[(games_df_3["rank_total"].isna()) | (games_df_3["rank_total"] <= 1), "rank_norm"] = 0
games_df_3["rank_pos"] = games_df_3["rank_pos"].fillna(0)
games_df_3["rank_norm"] = games_df_3["rank_norm"].fillna(0)
games_df_3["rank_total"] = games_df_3["rank_total"].fillna(0)
# -------------------------
# Price buckets (recommended for content-based)
# -------------------------
games_df_3["price_bucket"] = games_df_3[COL_PRICE].apply(price_bucket)

# -------------------------
# Build combined text field
# -------------------------
games_df_3["text"] = (
    games_df_3[COL_APP].apply(clean_text) + " " +
    games_df_3[COL_CAT].apply(clean_text) + " " +
    games_df_3[COL_GENRE].apply(clean_text) + " " +
    games_df_3[COL_DESC].apply(clean_text) + " " +
    games_df_3[COL_PUBLISHER].apply(clean_text) + " " +
    games_df_3[COL_ISRANKED].apply(clean_text) + " " +
    games_df_3["price_bucket"].apply(clean_text)
)

# Quick check
# games_df_3[["name", "price", "price_bucket", "rank", "rank_pos", "rank_total", "rank_norm", "text"]].head()

display(games_df_3.head())
display(games_df_3.info())

,rank type,name,rankScore,sinopsis,tags,rank,price,Category,Published by,rank_pos,rank_total,rank_norm,price_bucket,text
0,ranked,A Short Hike,9.34,A Short Hike is an adventure game for the Nint...,"Tags - Adventure, Exploration, Open world, Col...",16 / 4512,6.29,Adventure,Whippoorwill,16.0,4512.0,0.996675,budget,"a short hike adventure tags - adventure, explo..."
1,ranked,Kentucky Route Zero: TV Edition,9.10,Kentucky Route Zero: TV Edition is an adventur...,"Tags - Adventure, Story choices, Emotional, Me...",49 / 4512,20.99,Adventure,Annapurna Interactive,49.0,4512.0,0.989359,mid,kentucky route zero: tv edition adventure tags...
2,ranked,The Oregon Trail,9.07,The Oregon Trail is an adventure game for the ...,"Tags - Adventure, Survival, Story choices, Wes...",54 / 4512,26.99,Adventure,Gameloft,54.0,4512.0,0.988251,mid,"the oregon trail adventure tags - adventure, s..."
3,ranked,13 Sentinels: Aegis Rim,8.92,13 Sentinels: Aegis Rim is an adventure game f...,"Tags - Adventure, RPG, Strategy, Visual novel,...",97 / 4512,49.99,Adventure,Atlus Sega,97.0,4512.0,0.978719,premium,13 sentinels: aegis rim adventure tags - adven...
4,ranked,AI: The Somnium Files,8.86,AI: The Somnium Files is an adventure game for...,"Tags - Adventure, Point-and-click, Visual nove...",116 / 4512,35.99,Adventure,Spike Chunsoft,116.0,4512.0,0.974507,premium,ai: the somnium files adventure tags - adventu...


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6820 entries, 0 to 6819
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rank type     6820 non-null   object 
 1   name          6820 non-null   object 
 2   rankScore     6820 non-null   float64
 3   sinopsis      6820 non-null   object 
 4   tags          6820 non-null   object 
 5   rank          6820 non-null   object 
 6   price         6820 non-null   float64
 7   Category      6820 non-null   object 
 8   Published by  6820 non-null   object 
 9   rank_pos      6820 non-null   float64
 10  rank_total    6820 non-null   float64
 11  rank_norm     6820 non-null   float64
 12  price_bucket  6820 non-null   object 
 13  text          6820 non-null   object 
dtypes: float64(5), object(9)
memory usage: 746.1+ KB


None

This code prepares the dataset for a content-based recommendation system.
It first defines and validates required columns, ensuring important fields like name and tags exist.
Then it cleans and standardizes text fields by handling missing values, converting to strings, and normalizing formatting.
It removes duplicate games based on their name to avoid repeated items in the recommendation results.
The code also extracts structured information from the rank column and creates a price bucket feature from the numeric price.
Finally, it combines multiple fields (name, category, tags, synopsis, publisher, rank type, and price bucket) into a single text column that will be used for similarity computation.

In [ ]:
# Helper: parse genres/tags for evaluation
import ast
import re

def parse_tags(x):
    if pd.isna(x):
        return set()

    x = str(x).strip()

    if x == "no tags":
        return set()

    # Case 1: looks like a Python list stored as string
    if x.startswith("[") and x.endswith("]"):
        try:
            parsed = ast.literal_eval(x)
            return set(str(i).strip().lower() for i in parsed if str(i).strip())
        except:
            pass

    # Case 2: normal delimiter-based string
    parts = re.split(r"[;,|/]", x)
    parts = [p.strip().lower() for p in parts if p.strip()]
    return set(parts)

games_df_4 = games_df_3.copy()
games_df_4["tags"] = games_df_4[COL_GENRE].apply(parse_tags)
games_df_4[["tags"]].head()

,tags
0,"{pixel art, collect-a-thon, dogs, family frien..."
1,"{story choices, emotional, melancholy, tags - ..."
2,"{80's retrogaming, survival, western, story ch..."
3,"{visual novel, mecha, science fiction, robots,..."
4,"{visual novel, point-and-click, science fictio..."


This step extracts and processes genre information into structured tag sets for each application. The tags are cleaned and split based on common separators to enable comparison between items. These tag sets are later used to evaluate recommendation quality using genre overlap as a relevance proxy.

In [ ]:
# Preprocessed dataset
display(games_df_4.info())
display(games_df_4.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6820 entries, 0 to 6819
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rank type     6820 non-null   object 
 1   name          6820 non-null   object 
 2   rankScore     6820 non-null   float64
 3   sinopsis      6820 non-null   object 
 4   tags          6820 non-null   object 
 5   rank          6820 non-null   object 
 6   price         6820 non-null   float64
 7   Category      6820 non-null   object 
 8   Published by  6820 non-null   object 
 9   rank_pos      6820 non-null   float64
 10  rank_total    6820 non-null   float64
 11  rank_norm     6820 non-null   float64
 12  price_bucket  6820 non-null   object 
 13  text          6820 non-null   object 
dtypes: float64(5), object(9)
memory usage: 746.1+ KB


None

(6820, 14)

In [ ]:
# === MODEL 1 — TF-IDF + Cosine Similarity ===

# Build TF-IDF matrix
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf = TfidfVectorizer(
    ngram_range=(1,2),
    stop_words="english",
    min_df=2,
    max_df=0.9
)

X_tfidf = tfidf.fit_transform(games_df_4["text"])
display(X_tfidf.shape)
display(len(tfidf.vocabulary_))

(6820, 10806)

10806

This step transforms the combined text data into numerical feature vectors using the TF-IDF (Term Frequency–Inverse Document Frequency) method. TF-IDF represents each application based on the importance of words in its textual content.

The TF-IDF matrix shape (13145, 17027) means that 13,145 games were transformed into vectors with 17027 features each. Each row represents one game, and each column represents a word or two-word phrase (ngram). The number 17027 confirms that the vocabulary size is exactly 17027.

min_df=2 removes words that appear in fewer than 2 documents, filtering out very rare or noisy terms that don’t help generalize similarity.

max_df=0.9 removes words that appear in more than 90% of documents, eliminating overly common terms that don’t help distinguish between games.

In [ ]:
# Recommend Top-N similar apps (TF-IDF)
import numpy as np
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

games_df_5 = games_df_4.copy()

# Map app name -> index (after reset_index, this matches X_tfidf row order)
games_df_5 = games_df_5.reset_index(drop=True)
name_to_idx = pd.Series(games_df_5.index, index=games_df_5[COL_APP]).to_dict()

# Choose which columns to show in recommendation output
OUTPUT_COLS = [
    COL_APP,        # name
    COL_GENRE,      # tags
    COL_DESC,       # sinopsis
    COL_CAT,        # Category
    COL_PRICE,      # price
    COL_RATING,     # rankScore
    COL_RANK,       # rank (e.g., "16 / 4512")
    COL_ISRANKED,   # rank type
    COL_PUBLISHER   # Published by
]

# Keep only columns that actually exist (prevents KeyError)
OUTPUT_COLS = [c for c in OUTPUT_COLS if c in games_df_5.columns]

def recommend_tfidf(app_name: str, top_n: int = 10, return_cols=None):
    if app_name not in name_to_idx:
        # helpful suggestion: show close matches
        matches = games_df_5[COL_APP].str.contains(app_name, case=False, na=False)
        suggestions = games_df_5.loc[matches, COL_APP].head(10).tolist()
        raise ValueError(f"App name not found. Try exact match. Suggestions: {suggestions}")

    idx = name_to_idx[app_name]

    # Similarity of this game against all games
    sims = cosine_similarity(X_tfidf[idx], X_tfidf).ravel()
    sims[idx] = -1  # exclude itself

    top_idx = np.argsort(sims)[::-1][:top_n]

    cols = OUTPUT_COLS if return_cols is None else return_cols
    cols = [c for c in cols if c in games_df_5.columns]

    recs = games_df_5.loc[top_idx, cols].copy()
    recs["score"] = sims[top_idx]

    # optional: nicer ordering (highest score first)
    recs = recs.sort_values("score", ascending=False).reset_index(drop=True)
    return recs

# Example
recommend_tfidf(games_df_5[COL_APP].iloc[0], top_n=10)

,name,tags,sinopsis,Category,price,rankScore,rank,rank type,Published by,score
0,Littlewood,"{no fail state, top-down, family friendly, cat...",Littlewood is a simulation for the Nintendo Sw...,Simulation,11.00,8.42,379 / 4512,ranked,Smash Games,0.326825
1,Alba: A Wildlife Adventure,"{no fail state, family friendly, real-time, wa...",Alba: A Wildlife Adventure is an adventure gam...,Adventure,16.74,7.70,1414 / 4512,ranked,Plug In Digital,0.323918
2,Embracelet,"{no fail state, walking simulator, tags - adve...",Embracelet is an interactive story for the Nin...,Adventure » Interactive story,9.99,8.56,256 / 4512,ranked,Machineboy,0.279439
3,Yoku's Island Express,"{no fail state, family friendly, feel good, ca...",Yoku's Island Express is a metroidvania for th...,Metroidvania,15.99,8.69,165 / 4512,ranked,Team17,0.265780
4,Celeste,"{no fail state, accessibility options, feel go...",Celeste is a platformer (2d) for the Nintendo ...,Platformer » Platformer (2D),17.99,9.60,4 / 4512,ranked,Matt Makes Games,0.264832
5,AER Memories of Old,"{no fail state, family friendly, tags - advent...",AER Memories of Old is an interactive story fo...,Adventure » Interactive story,8.00,7.53,1735 / 4512,ranked,Daedalic Entertainment,0.263457
6,Serene Hike,{tags - walking simulator},Serene Hike is an interactive story for the Ni...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.256214
7,Luigi's Mansion 3,"{family friendly, cats, tags - action adventur...",Luigi's Mansion 3 is an action-adventure game ...,Adventure » Action-adventure,58.98,8.79,135 / 4512,ranked,Nintendo,0.252200
8,The Gardens Between,"{family friendly, tags - adventure, casual, lo...",The Gardens Between is a puzzle adventure game...,Adventure » Puzzle adventure,17.99,8.53,277 / 4512,ranked,The Voxel Agents,0.248532
9,Woodland Hike,"{tags - interactive story, walking simulator}",Woodland Hike is an interactive story for the ...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.240547


This step computes cosine similarity between the selected app and all other apps using the TF-IDF feature matrix. The apps are then ranked based on similarity scores. Finally, the top-N most similar applications are returned as recommendations.

In [ ]:
# === MODEL 2 — Sentence-BERT Embeddings + Cosine Similarity ===

from sentence_transformers import SentenceTransformer
import numpy as np

games_df_6 = games_df_4.copy().reset_index(drop=True)

# Load model (fast + good balance)
sbert_model = SentenceTransformer("all-MiniLM-L6-v2")

# Encode text column
embeddings = sbert_model.encode(
    games_df_6["text"].tolist(),
    normalize_embeddings=True,      # important for cosine similarity
    show_progress_bar=True
)

embeddings.shape

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Batches:   0%|          | 0/214 [00:00<?, ?it/s]

(6820, 384)

This step converts the combined text of each application into dense vector representations using a pre-trained Sentence-BERT model. These embeddings capture semantic meaning beyond simple word frequency. The resulting vectors are used to compute similarity between apps at a deeper contextual level.

In [ ]:
# Recommend Top-N similar apps (Embeddings)
import numpy as np

# Make sure index alignment matches embeddings
games_df_6 = games_df_4.copy().reset_index(drop=True)

# Map name -> index (aligned with embeddings)
name_to_idx_sbert = pd.Series(
    games_df_6.index,
    index=games_df_6[COL_APP]
).to_dict()

# Columns to return (reuse your OUTPUT_COLS logic if defined)
OUTPUT_COLS_SBERT = [
    COL_APP,
    COL_GENRE,
    COL_DESC,
    COL_CAT,
    COL_PRICE,
    COL_RATING,
    COL_RANK,
    COL_ISRANKED,
    COL_PUBLISHER
]

OUTPUT_COLS_SBERT = [c for c in OUTPUT_COLS_SBERT if c in games_df_6.columns]


def recommend_sbert(app_name: str, top_n: int = 10):
    if app_name not in name_to_idx_sbert:
        matches = games_df_6[COL_APP].str.contains(app_name, case=False, na=False)
        suggestions = games_df_6.loc[matches, COL_APP].head(10).tolist()
        raise ValueError(f"App name not found. Suggestions: {suggestions}")

    idx = name_to_idx_sbert[app_name]

    # Since embeddings were normalized, dot product = cosine similarity
    sims = embeddings @ embeddings[idx]

    sims[idx] = -1  # exclude itself

    top_idx = np.argsort(sims)[::-1][:top_n]

    recs = games_df_6.loc[top_idx, OUTPUT_COLS_SBERT].copy()
    recs["score"] = sims[top_idx]

    return recs.sort_values("score", ascending=False).reset_index(drop=True)


# Example
recommend_sbert(games_df_6[COL_APP].iloc[0], top_n=10)

,name,tags,sinopsis,Category,price,rankScore,rank,rank type,Published by,score
0,Hike Trip,{tags - walking simulator},Hike Trip is an interactive story for the Nint...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.768967
1,Hike,"{tags - interactive story, walking simulator}",Hike is an interactive story for the Nintendo ...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.734441
2,Hike Valley,{tags - walking simulator},Hike Valley is an interactive story for the Ni...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.718233
3,Grand Hike,{tags - walking simulator},Grand Hike is an interactive story for the Nin...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.711555
4,Fishing Vacation,"{top-down, tags - adventure, simulation, fishi...",Fishing Vacation is an adventure game for the ...,Adventure,1.69,7.50,1769 / 4512,ranked,Dread XP,0.686849
5,Lonely Mountains: Downhill,"{stunts, family friendly, top-down, racing, sp...",Lonely Mountains: Downhill is a sports game fo...,Sports,17.99,8.44,363 / 4512,ranked,Thunderful,0.686203
6,Haven Park,"{family friendly, top-down, walking simulator,...",Haven Park is an adventure game for the Ninten...,Adventure,6.99,7.92,1021 / 4512,ranked,Mooneye Studios,0.685380
7,Loop,"{top-down, puzzle platformer, tags - adventure...",Loop is a puzzle adventure game for the Ninten...,Adventure » Puzzle adventure,5.00,0.00,TBC,unranked,Team Chili,0.675982
8,Work Trip,"{tags - interactive story, walking simulator, ...",Work Trip is an interactive story for the Nint...,Adventure » Interactive story,4.49,0.00,TBC,unranked,Ultimate Games,0.675469
9,Button City,"{family friendly, cats, tags - adventure, feel...",Button City is an adventure game for the Ninte...,Adventure,15.99,7.07,2502 / 4512,ranked,Subliminal,0.673410


This step calculates cosine similarity between the selected app’s embedding vector and all other app embeddings. The apps are then ranked according to their semantic similarity scores. The top-N most similar applications are returned as recommendations.

In [ ]:
# Because we don’t have user interactions, we define “relevant” as:
#   Recommended game shares at least 1 genre tag with the query game.

# Offline Evaluation (Precision@K with Genre Overlap Proxy)
def precision_at_k_from_names(rec_names, query_tags, df_ref, name_to_idx_ref, k=10):
    rec_names = rec_names[:k]
    if len(rec_names) == 0:
        return 0.0

    hits = 0
    for name in rec_names:
        j = name_to_idx_ref.get(name, None)
        if j is None:
            continue

        rec_tags = df_ref.loc[j, "tags"]
        if len(query_tags.intersection(rec_tags)) > 0:
            hits += 1

    return hits / k

def eval_precision_model(
    recommender_fn,
    df_ref,
    name_to_idx_ref,
    sample_size=300,
    k=10,
    seed=42
):
    rng = np.random.default_rng(seed)
    idxs = rng.choice(len(df_ref), size=min(sample_size, len(df_ref)), replace=False)

    scores = []
    for idx in idxs:
        app = df_ref.loc[idx, COL_APP]
        query_tags = df_ref.loc[idx, "tags"]

        if len(query_tags) == 0:
            continue

        rec_df = recommender_fn(app, top_n=k)
        rec_names = rec_df[COL_APP].tolist()

        p = precision_at_k_from_names(
            rec_names,
            query_tags,
            df_ref,
            name_to_idx_ref,
            k=k
        )

        scores.append(p)

    return float(np.mean(scores)) if scores else 0.0

p_tfidf = eval_precision_model(
    recommend_tfidf,
    games_df_5,
    name_to_idx,
    sample_size=300,
    k=10
)

p_sbert = eval_precision_model(
    recommend_sbert,
    games_df_6,
    name_to_idx_sbert,
    sample_size=300,
    k=10
)

p_tfidf, p_sbert


(0.9056666666666668, 0.8426666666666667)

This code evaluates the recommendation models using an offline metric called Precision@K. Since we do not have real user interaction data, a recommendation is considered “relevant” if it shares at least one genre tag with the query game. The precision_at_k_from_names function computes how many of the Top-K recommended games overlap in genre with the query game. The eval_precision_model function samples random games, generates recommendations, and calculates the average Precision@K score. This provides a quantitative way to compare TF-IDF and SBERT models. The final results indicate that SBERT slightly outperforms TF-IDF in genre-based relevance under this proxy evaluation.

In [ ]:
# Compare results + inspect outputs
print(f"TF-IDF Precision@10 (genre overlap proxy): {p_tfidf:.4f}")
print(f"SBERT  Precision@10 (genre overlap proxy): {p_sbert:.4f}")

query_app = "Xenoblade Chronicles 3"
print("\nQuery:", query_app)

# TF-IDF (Model 1)
print("\nTF-IDF Recommendations:")
try:
    display(recommend_tfidf(query_app, top_n=10))
except ValueError as e:
    print("TF-IDF:", e)
    print("Example names:", games_df_5[COL_APP].head(20).tolist())

# SBERT (Model 2)
print("\nSBERT Recommendations:")
try:
    display(recommend_sbert(query_app, top_n=10))
except ValueError as e:
    print("SBERT:", e)
    print("Example names:", games_df_6[COL_APP].head(20).tolist())

TF-IDF Precision@10 (genre overlap proxy): 0.9057
SBERT  Precision@10 (genre overlap proxy): 0.8427

Query: Xenoblade Chronicles 3

TF-IDF Recommendations:


,name,tags,sinopsis,Category,price,rankScore,rank,rank type,Published by,score
0,Xenoblade Chronicles 2,"{science fiction, real-time, third person, exp...",Xenoblade Chronicles 2 is an action rpg for th...,RPG » Action RPG,76.98,8.40,389 / 4512,ranked,Nintendo,0.853980
1,Xenoblade Chronicles: Definitive Edition,"{anime, science fiction, real-time, tags - exp...",Xenoblade Chronicles: Definitive Edition is an...,RPG » Action RPG,49.99,8.97,86 / 4512,ranked,Nintendo,0.734297
2,.hack//G.U. Last Recode,"{00's retrogaming, anime, science fiction, rea...",.hack//G.U. Last Recode is an action rpg for t...,RPG » Action RPG,44.99,7.33,2069 / 4512,ranked,Bandai Namco Entertainment,0.363891
3,Ravensword: Shadowlands,"{fantasy, third person, real-time, tags - rpg}",Ravensword: Shadowlands is an action rpg for t...,RPG » Action RPG,6.99,0.00,TBC,unranked,Ratalaika Games,0.340748
4,Undungeon,"{fantasy, hack & slash, top-down, exploration,...",Undungeon is an action rpg for the Nintendo Sw...,RPG » Action RPG,17.99,5.66,4056 / 4512,ranked,tinyBuild Games,0.310833
5,Hatchwell,"{fantasy, hack & slash, top-down, exploration,...",Hatchwell is an action rpg for the Nintendo Sw...,RPG » Action RPG,7.39,0.00,TBC,unranked,Adrian Corpuz,0.306814
6,Dragon's Dogma: Dark Arisen,"{fantasy, real-time, tags - adventure, third p...",Dragon's Dogma: Dark Arisen is an action rpg f...,RPG » Action RPG,24.99,8.45,355 / 4512,ranked,Capcom,0.305180
7,Eternal Radiance,"{fantasy, anime, real-time, third person, tags...",Eternal Radiance is an action rpg for the Nint...,RPG » Action RPG,15.29,5.33,4181 / 4512,ranked,Visualnoveler,0.304833
8,The Elder Scrolls V: Skyrim,"{fantasy, real-time, tags - exploration, third...",The Elder Scrolls V: Skyrim is an action rpg f...,RPG » Action RPG,59.99,9.02,68 / 4512,ranked,Bethesda Softworks,0.302385
9,Crystar,"{fantasy, anime, tags - dungeon crawler, hack ...",Crystar is an action rpg for the Nintendo Swit...,RPG » Action RPG,44.99,6.05,3777 / 4512,ranked,NIS America,0.299045



SBERT Recommendations:


,name,tags,sinopsis,Category,price,rankScore,rank,rank type,Published by,score
0,Xenoblade Chronicles 2,"{science fiction, real-time, third person, exp...",Xenoblade Chronicles 2 is an action rpg for th...,RPG » Action RPG,76.98,8.40,389 / 4512,ranked,Nintendo,0.949090
1,Xenoblade Chronicles: Definitive Edition,"{anime, science fiction, real-time, tags - exp...",Xenoblade Chronicles: Definitive Edition is an...,RPG » Action RPG,49.99,8.97,86 / 4512,ranked,Nintendo,0.861304
2,Dragon Ball Xenoverse 2,"{fighting, anime, third person, tags - rpg}",Dragon Ball Xenoverse 2 is an action rpg for t...,RPG » Action RPG,17.99,7.74,1367 / 4512,ranked,Bandai Namco Entertainment,0.771186
3,Metal Max Xeno Reborn,"{anime, third person, tags - rpg, exploration,...",Metal Max Xeno Reborn is an RPG for the Ninten...,RPG,34.99,7.27,2193 / 4512,ranked,PQube,0.747950
4,Crystar,"{fantasy, anime, tags - dungeon crawler, hack ...",Crystar is an action rpg for the Nintendo Swit...,RPG » Action RPG,44.99,6.05,3777 / 4512,ranked,NIS America,0.656374
5,Warframe,"{space, crafting, shooting, science fiction, t...",Warframe is a third-person shooter for the Nin...,Shooting » Third-person shooter,0.00,8.59,239 / 4512,ranked,Digital Extremes,0.648215
6,Fire Emblem Warriors: Three Hopes,"{fantasy, anime, third person, tags - hack & s...",Fire Emblem Warriors: Three Hopes is an action...,RPG » Action RPG,49.99,8.30,505 / 4512,ranked,Nintendo,0.643918
7,Daemon X Machina,"{anime, shooting, science fiction, robots, tag...",Daemon X Machina is a shooting game for the Ni...,Shooting,49.99,7.16,2378 / 4512,ranked,Nintendo,0.629173
8,Minecraft Legends,"{fantasy, tags - dungeon crawler, hack & slash...",Minecraft Legends is an action rpg for the Nin...,RPG » Action RPG,41.74,6.88,2820 / 4512,ranked,Mojang,0.625568
9,Earthen Dragon,"{fantasy, hack & slash, tags - action adventur...",Earthen Dragon is an action rpg for the Ninten...,RPG » Action RPG,8.89,0.00,TBC,unranked,Origamihero Games,0.621905


This step compares the performance of the two recommendation models using the Precision@10 metric and displays the results in a readable format. It prints the average Precision@10 scores for both the TF-IDF and SBERT models, allowing a direct quantitative comparison. Then, it selects a specific query game to qualitatively inspect the recommendation outputs. The system generates and displays the Top-10 recommendations from both models for the same query. This helps visually evaluate how similar and relevant the recommended games are beyond the numerical metric.